test/test-template-engine.js

```diff
@@ -53,13 +53,66 @@ describe('html rendering', () => {
     assert(container.querySelector('#target').textContent === 'No interpolation.');
   });
 
+  it('renders comments', () => {
+    const container = document.createElement('div');
+    render(container, html`<!--This is HTML: "&ldquo;<div></div>&rdquo;"-->`);
+    assert(container.childNodes.length === 1);
+    // Note that character references are _not_ replaced — this is _just text_.
+    assert(container.childNodes[0].textContent === 'This is HTML: "&ldquo;<div></div>&rdquo;"');
+  });
+
+  it('renders template elements', () => {
+    // It’s important that the _content_ is populated here. Not the template.
+    const container = document.createElement('div');
+    render(container, html`<template><div><p></p></div></template>`);
+    assert(!!container.querySelector('template'));
+    assert(container.querySelector('template').content.childElementCount === 1);
+    assert(container.querySelector('template').content.children[0].childElementCount === 1);
+    assert(container.querySelector('template').content.children[0].children[0].localName === 'p');
+  });
+
+  it('renders pre elements with optional, initial newline', () => {
+    const expected = '        hi\n      ';
+    const container = document.createElement('div');
+    render(container, html`
+      <pre>
+        <span>hi</span>
+      </pre>
+    `);
+    assert(container.querySelector('pre').textContent === expected); // first newline is removed
+  });
+
   it('renders named html entities in text content', () => {
     const container = document.createElement('div');
     render(container, html`<div>${'--'}&#123;&lt;&amp;&gt;&apos;&quot;&#x007D;${'--'}</div>`);
     assert(container.childNodes.length === 1);
     assert(container.childNodes[0].textContent === `--{<&>'"}--`);
   });
 
+  it('renders direct references in replaceable character data', () => {
```

Comment:
Some interesting DOM-specific test cases are moved from the parser tests to the template engine tests because the _require rendering_ to test for correctness.

For example, we can test that character references are _interpreted_ in the parser, but we must ensure that they are then properly _decoded_ in the template engine (which requires DOM APIs).